presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankAccumulator.java

```diff
@@ -0,0 +1,894 @@
+
+            peerGroupIndex = peerGroupBuffer.getNextPeerIndex(peerGroupIndex);
+        }
+        while (peerGroupIndex != UNKNOWN_INDEX);
+        verify(actualPeerGroupCount == peerGroupCount, "Recorded peer group count does not match actual");
+
+        if (leftChildHeapIndex != UNKNOWN_INDEX) {
+            verify(rowComparisonStrategy.compare(rowId, peerGroupBuffer.getRowId(heapNodeBuffer.getPeerGroupIndex(leftChildHeapIndex))) > 0, "Max heap invariant violated");
+        }
+        if (rightChildHeapIndex != UNKNOWN_INDEX) {
+            verify(leftChildHeapIndex != UNKNOWN_INDEX, "Left should always be inserted before right");
+            verify(rowComparisonStrategy.compare(rowId, peerGroupBuffer.getRowId(heapNodeBuffer.getPeerGroupIndex(rightChildHeapIndex))) > 0, "Max heap invariant violated");
+        }
+
+        IntegrityStats leftIntegrityStats = verifyHeapIntegrity(groupId, leftChildHeapIndex);
+        IntegrityStats rightIntegrityStats = verifyHeapIntegrity(groupId, rightChildHeapIndex);
+
+        verify(Math.abs(leftIntegrityStats.getMaxDepth() - rightIntegrityStats.getMaxDepth()) <= 1, "Heap not balanced");
+
+        return new IntegrityStats(
+                Math.max(leftIntegrityStats.getMaxDepth(), rightIntegrityStats.getMaxDepth()) + 1,
+                leftIntegrityStats.getPeerGroupCount() + rightIntegrityStats.getPeerGroupCount() + 1,
+                leftIntegrityStats.getValueCount() + rightIntegrityStats.getValueCount() + peerGroupCount);
+    }
+
+    private static class IntegrityStats
+    {
+        private final long maxDepth;
+        private final long peerGroupCount;
+        private final long valueCount;
+
+        public IntegrityStats(long maxDepth, long peerGroupCount, long valueCount)
+        {
+            this.maxDepth = maxDepth;
+            this.peerGroupCount = peerGroupCount;
+            this.valueCount = valueCount;
+        }
+
+        public long getMaxDepth()
+        {
+            return maxDepth;
+        }
+
+        public long getPeerGroupCount()
+        {
+            return peerGroupCount;
+        }
+
+        public long getValueCount()
+        {
+            return valueCount;
+        }
+    }
+
+    /**
+     * Buffer abstracting a mapping from group ID to a heap. The group ID provides the index for all operations.
+     */
+    private static class GroupIdToHeapBuffer
+    {
+        private static final long INSTANCE_SIZE = ClassLayout.parseClass(GroupIdToHeapBuffer.class).instanceSize();
+
+        /*
+         *  Memory layout:
+         *  [LONG] heapNodeIndex1,
+         *  [LONG] heapNodeIndex2,
+         *  ...
+         */
+        // Since we have a single element per group, this array is effectively indexed on group ID
+        private final LongBigArray heapIndexBuffer = new LongBigArray(UNKNOWN_INDEX);
+
+        /*
+         *  Memory layout:
+         *  [LONG] valueCount1, [LONG] heapSize1,
```

Comment:
We could consider `int` here. Do we expect queries to actually limit rank/row number above 2^32? This could save a lot of mem.

Reply:
Yes big savings here if we can limit to Integer size. I put it this way because the original Grouped TopN for row number implementation seemed to allow LONG output sizes for row number. But other than that, I don't see a real problem capping max rank at 32 bits.

Reply:
@sopel39, also if we change this, we have to change some of the other metrics stored in other data structures to also be int to match. That would cause additional cache misses to pull them into a different data structure. I think it will be fine either way, but I know you care about that, so just want to check what your preference is here. I don't mind either way since these are not the main bottlenecks right now.

Reply:
ping @sopel39 on this one